scripts/build_powerplants.py

```diff
@@ -35,71 +36,68 @@
-    return cntry.alpha_2
 
 if __name__ == "__main__":
     if 'snakemake' not in globals():
         from vresutils.snakemake import MockSnakemake, Dict
 
         snakemake = MockSnakemake(
-            input=Dict(base_network='networks/base.nc'),
+            input=Dict(base_network='networks/base.nc',
+                       custom_powerplants='data/custom_powerplants.csv'),
             output=['resources/powerplants.csv']
         )
 
     logging.basicConfig(level=snakemake.config['logging_level'])
 
     n = pypsa.Network(snakemake.input.base_network)
+    countries = n.buses.country.unique()
 
-    ppm.powerplants(from_url=True)
-
-    ppl = (ppm.collection.matched_data()
-        [lambda df : ~df.Fueltype.isin(('Solar', 'Wind'))]
-        .pipe(ppm.cleaning.clean_technology)
-        .assign(Fueltype=lambda df: (
-            df.Fueltype.where(df.Fueltype != 'Natural Gas',
-                                df.Technology.replace('Steam Turbine', 'OCGT').fillna('OCGT'))))
-        .pipe(ppm.utils.fill_geoposition))
+    ppl = (pm.powerplants(from_url=True)
+           .powerplant.convert_country_to_alpha2()
+           .query('Fueltype not in ["Solar", "Wind"] and Country in @countries')
+           .replace({'Technology': {'Steam Turbine': 'OCGT'}})
+            .assign(Fueltype=lambda df: (
+                    df.Fueltype
+                      .where(df.Fueltype != 'Natural Gas',
+                             df.Technology.replace('Steam Turbine',
+                                                   'OCGT').fillna('OCGT')))))
 
-    # ppl.loc[(ppl.Fueltype == 'Other') & ppl.Technology.str.contains('CCGT'), 'Fueltype'] = 'CCGT'
-    # ppl.loc[(ppl.Fueltype == 'Other') & ppl.Technology.str.contains('Steam Turbine'), 'Fueltype'] = 'CCGT'
+    ppl_query = snakemake.config['electricity']['powerplants_filter']
+    if isinstance(ppl_query, str):
+        ppl.query(ppl_query, inplace=True)
 
-    ppl = ppl.loc[ppl.lon.notnull() & ppl.lat.notnull()]
-    ppl = ppl.replace({"Country": {"Macedonia, Republic of": "North Macedonia"}}) 
+    ppl = add_custom_powerplants(ppl) # add carriers from own powerplant files
 
-    ppl_country = ppl.Country.map(country_alpha_2)
-    countries = n.buses.country.unique()
-    cntries_without_ppl = []
+    cntries_without_ppl = [c for c in countries if c not in ppl.Country.unique()]
 
-    for cntry in countries:
-        substation_lv_i = n.buses.index[n.buses['substation_lv'] & (n.buses.country == cntry)]
-        ppl_b = ppl_country == cntry
-        if not ppl_b.any():
-            cntries_without_ppl.append(cntry)
-            continue
+    substation_i = n.buses.query('substation_lv').index
+    kdtree = KDTree(n.buses.loc[substation_i, ['x','y']].values)
 
-        kdtree = KDTree(n.buses.loc[substation_lv_i, ['x','y']].values)
-        ppl.loc[ppl_b, 'bus'] = substation_lv_i[kdtree.query(ppl.loc[ppl_b, ['lon','lat']].values)[1]]
+    ppl['bus'] = substation_i[kdtree.query(ppl[['lon','lat']].values)[1]]
```

Comment:
I tried here to not mistakenly assign a powerplant to a neighbouring bus in another country.

Can you output a warning of how many and maybe which powerplants, are now assigned to the "wrong" country!

Reply:
yes, good point. I did not think about it. I think I'll reverse that.